src/app/scanning.service.ts

```diff
@@ -0,0 +1,167 @@
+    return this.children;
+  }
+
+  getChild(index: number): Scannable {
+    return this.children[index];
+  }
+
+  addChild(child: Scannable) {
+    this.children.push(child);
+  }
+
+  sortChildren() {
+    this.children.sort(this._sortScannables);
+  }
+
+  _sortScannables(a: Scannable, b: Scannable): number {
+    return a.priority - b.priority;
+  }
+}
+
+export class ScanningModel {
+  currentHighlight: Scannable;
+  currentSelection: Scannable;
+}
+
+export interface ScannableCollectionProvider extends Subscriber<ScanningModel> {
+  getScannableCollections(): ScannableCollection[];
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ScanningService {
+
+  private topLevelScannables: ScannableCollection;
+  private currentCollection: ScannableCollection;
+  private scanningModel: ScanningModel;
+  private currentSelectedIndex = 0;
+  private observers: ScannableCollectionProvider[];
+  private intervalId;
+
+  constructor(private configService: ConfigService) {
+    this.topLevelScannables = new ScannableCollection(0, '');
+    this.currentCollection = this.topLevelScannables;
+    this.scanningModel = new ScanningModel();
+    this.observers = [];
+  }
+
+  getScanningModel(): Observable<ScanningModel> {
+    return new Observable<ScanningModel>(this.registerScannable);
+  }
+
+  registerScannable = (observer: ScannableCollectionProvider): Subscription => {
+    this.observers.push(observer);
+
+    observer.getScannableCollections().forEach(col => col.sortChildren());
+
+    this.topLevelScannables.getChildren().push(...observer.getScannableCollections());
+    this.topLevelScannables.sortChildren();
+    this.startScanning();
+
+    return new Subscription(() => this._unsubscribe(observer));
+  }
+
+  _unsubscribe(observer: ScannableCollectionProvider) {
+    this.observers.splice(this.observers.indexOf(observer), 1);
+    observer.getScannableCollections().forEach(scannable => {
+      const collection = <ScannableCollection> scannable;
+      this.topLevelScannables.getChildren().splice(this.topLevelScannables.getChildren().indexOf(collection), 1);
+    });
+
+    this.currentCollection = this.topLevelScannables;
+
+    if (this.observers.length === 0) {
+      clearInterval(this.intervalId);
+      this.intervalId = undefined;
+      this.scanningModel = new ScanningModel();
+    }
+  }
+
+  startScanning() {
+    this.currentSelectedIndex = 0;
+
+    if (this.intervalId === undefined && this.observers.length >= 1 && this.configService.scanningConfig.enabled) {
+      this.intervalId = setInterval(this.updateHighlighted, this.configService.scanningConfig.time);
+    }
+  }
+
+  updateHighlighted = () => {
+    if (this.currentSelectedIndex >= this.currentCollection.getChildren().length) {
+      this.currentSelectedIndex = 0;
+    }
+
+    this.scanningModel.currentHighlight = this.currentCollection.getChild(this.currentSelectedIndex++);
+    this.observers.forEach(observer => observer.next(this.scanningModel));
```

Comment:
I think this might be better contained in a notifyObservers method. I know it's just one line, but then it would have a name and it just makes it quicker to work out what is happening

Reply:
Multicasting at it's finest though <3